src/presetsController.ts

```diff
@@ -0,0 +1,844 @@
+      return this.handleNoConfigurePresets();
+    }
+
+    interface AddPresetQuickPickItem extends vscode.QuickPickItem {
+      name: string;
+    }
+
+    enum SpecialOptions {
+      CreateFromConfigurationPreset = '__createFromConfigurationPreset__',
+      InheritBuildPreset = '__inheritBuildPreset__',
+      Custom = '__custom__'
+    }
+
+    const items: AddPresetQuickPickItem[] = [{
+        name: SpecialOptions.CreateFromConfigurationPreset,
+        label: localize('create.build.from.config.preset', 'Create from Configure Preset'),
+        description: localize('description.create.build.from.config.preset', 'Create a new build preset')
+    }];
+    if (preset.buildPresets().length > 0) {
+      items.push({
+        name: SpecialOptions.InheritBuildPreset,
+        label: localize('inherit.build.preset', 'Inherit from Build Preset'),
+        description: localize('description.inherit.build.preset', 'Inherit from an existing build preset')
+      });
+    }
+    items.push({
+      name: SpecialOptions.Custom,
+      label: localize('custom.build.preset', 'Custom'),
+      description: localize('description.custom.build.preset', 'Add an custom build preset')
+    });
+
+    const chosenItem = await vscode.window.showQuickPick(items,
+      { placeHolder: localize('add.a.build.preset.placeholder', 'Add a build preset for {0}', this.folder.name) });
+    if (!chosenItem) {
+      log.debug(localize('user.cancelled.add.build.preset', 'User cancelled adding build preset'));
+      return false;
+    } else {
+      let newPreset: preset.BuildPreset | undefined;
+      switch(chosenItem.name) {
+        case SpecialOptions.CreateFromConfigurationPreset: {
+          const placeHolder = localize('select.a.config.preset.placeholder', 'Select a configure preset');
+          const configurePreset = await this.showPresetSelector(preset.configurePresets(), { placeHolder });
+          newPreset = { name: '__placeholder__', description: '', displayName: '', configurePreset };
+          break;
+        }
+        case SpecialOptions.InheritBuildPreset: {
+          const placeHolder = localize('select.one.or.more.build.preset.placeholder', 'Select one or more build presets');
+          const inherits = await this.showPresetSelector(preset.buildPresets(), { placeHolder, canPickMany: true });
+          newPreset = { name: '__placeholder__', description: '', displayName: '', inherits };
+          break;
+        }
+        case SpecialOptions.Custom: {
+          newPreset = { name: '__placeholder__', description: '', displayName: '' };
+          break;
+        }
+        default:
+          break;
+      }
+
+      if (newPreset) {
+        const name = await this.showNameInputBox();
+        if (!name) {
+          return false;
+        }
+
+        newPreset.name = name;
+        await this.addPresetAddUpdate(newPreset, 'buildPresets');
+      }
+
+      return true;
+    }
+  }
+
+  async addTestPreset(): Promise<boolean> {
```

Comment:
Something is broken with this function.  I ran the command to add a test preset and got two entries added (neither of them looked correct):
```
        {
            "name": "__placeholder__",
            "description": "",
            "displayName": "",
            "configurePreset": "debug-x64"
        },
        {
            "name": "debug-x64-test",
            "description": "",
            "displayName": "",
            "configurePreset": "debug-x64"
        },
```

Reply:
Probably forgot something when moving the name input box to the end of the function. I'll take a look.